enhancements/update/synchronized-upgrades.md

```diff
@@ -0,0 +1,223 @@
+    upgrades, by failing its own upgrade, eg if an admin tries to
+    upgrade one cluster without the other, or tries to upgrade the two
+    clusters to different versions.
+
+    - (Or should it be possible to do z-stream upgrades of the tenant
+      cluster without bothering to upgrade the infra cluster too?)
+
+- The two clusters upgrade all of the operators up to MCO in parallel.
+
+- Whichever cluster reaches the MCO upgrade first needs to wait for
+  the other cluster to get there before proceeding. The two MCOs then
+  need to coordinate to complete the upgrade: First, they have to
+  agree on what order the physical hosts will be upgraded in. Second,
+  for each physical host, they have to properly synchronize the
+  upgrades of its infra node and its tenant node.
+
+  - More specifically, for each physical host, in some order:
+
+    - The Infra MCO will cordon and drain that host's infra node, and
+      the Tenant MCO will cordon and drain that host's tenant node.
+      (This can happen in parallel.)
+
+    - The Infra MCO will then upgrade the infra node (causing it to
+      reboot and temporarily break network connectivity to the tenant
+      node).
+
+    - Once the infra node upgrade completes, the Tenant MCO will
+      reboot and upgrade the tenant node.
+
+  - (This seems like it will absolutely require MCO changes.)
+
+  - One way to do this would be to have a CRD with an array of hosts,
+    indicating the ordering, and the current status of each host, and
+    the two MCOs could update and watch for updates to monitor each
+    other's progress.
+
+
+### Test Plan
+
+TBD
+
+### Graduation Criteria
+
+TBD
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+### Upgrade / Downgrade Strategy
+
+This is a modification to the upgrade process, not something that can
+be upgrade or downgraded on its own.
+
+TBD, as the details depend on the eventual design.
+
+### Version Skew Strategy
+
+TBD, as the details depend on the eventual design.
+
+We will need to deal with skew both within a single cluster, as well
+as skew between the infra and tenant clusters.
+
+### Operational Aspects of API Extensions
+
+TBD
+
+The only currently-proposed CRD is for Infra MCO to Tenant MCO
+communication, and would not be used by any other components.
+
+#### Failure Modes
+
+- The system might get confused and spuriously block upgrades that
+  should be allowed.
+
+- Communications failures might lead to upgrades failing without the
+  tenant cluster being able to figure out why they failed.
+
+- TBD
+
+#### Support Procedures
+
+TBD
+
+## Implementation History
+
+- Initial proposal: 2021-01-11
+
+## Drawbacks
+
+This makes the upgrade process more complicated, which risks rendering
+clusters un-upgradeable without manual intervention.
+
+However, without some form of synchronization, it is impossible to
+have non-disruptive tenant cluster upgrades.
+
+## Alternatives
```

Comment:
If the infra node workload only services the local tenant node then the infra cluster could update including staging machineconfig and RHCOS changes. Once that completes the tenant cluster updates and the MCO reboots tenant nodes which reboots the infra node into the updated RHCOS?

If an unexpected reboot seems like a concern perhaps the MCO on the infra node could be made to only apply the new desired configuration on a soft restart assuming rebooting the tenant triggers a soft reboot of the infra node?

Reply:
So as above, rebooting the tenant doesn't reboot the infra node, although I suspect there's probably some way we can detect when the host reboots.

So I guess with what you're suggesting, the only coordination that would be needed would be that you have to run the upgrade to almost-completion in the infra cluster, and then you start the upgrade in the tenant cluster, and as the tenant cluster upgrade completes it will also cause the completion of the infra cluster upgrade. Right? There'd still need to be (small) MCO changes to prevent the automatic reboot in the infra cluster.

Reply:
Yeah that's what I was thinking.

Reply:
It looks like it might not be possible for the BF-2 to directly detect a host reboot, at least when the card is in "isolated" mode.

But the infra node knows what pod-network pods are running on the tenant node, so it can infer that the tenant has been drained when (most of?) the pod-network pods go away. (If necessary, the dpu-operator could confirm this by looking at the Node object in the tenant cluster.) So we could say, if we have a reboot pending, and the tenant gets drained, then reboot. Even if we guessed wrong and the drain wasn't actually because of the MCO upgrade on the tenant, we still would at least avoid most of the disruption.

Reply:
Are there any files shared between the two?

Reply:
Reading a bit more doesn't look like there is.

Reply:
The two systems don't share a kernel (or a common CPU architecture even), so they can't share files.